finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +182,55 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn request(&mut self, hash: <B as BlockT>::Hash) {
+        if let Ok(Some(header)) = self.backend.blockchain().header(BlockId::Hash(hash)) {
+            let number = *header.number();
+            debug!(target: "aleph-justification", "Trying to request block {:?}", number);
+            self.request_status.save_block_number(number);
+            self.request_status.insert_hash(hash);
+            debug!(target: "aleph-justification", "We have block {:?} with hash {:?}. Requesting justification.", number, header.hash());
+            self.justification_request_scheduler.on_request_sent();
+            self.block_requester.request_justification(&hash, number);
+        } else {
+            debug!(target: "aleph-justification", "Cancelling request, because we don't have block {:?}.", hash);
+        }
+    }
+
+    // We request justifications for all the children of last finalized block and a justification
+    // for a block of number num on longest branch.
+    // Assuming that we request at the same pace that finalization is progressing, the former ensures
+    // that we are up to date with finalization. On the other hand, the former enables fast catch up
+    // if we are behind.
+    // We don't remove the child that it's on the same branch as best since a fork may happen
+    // somewhere in between them.
+    fn request_targets(&self, mut num: NumberFor<B>) -> Vec<<B as BlockT>::Hash> {
+        let blockchain_backend = self.backend.blockchain();
+        let blockchain_info = blockchain_backend.info();
+        let finalized_hash = blockchain_info.finalized_hash;
+
+        let mut targets = blockchain_backend
+            .children(finalized_hash)
+            .unwrap_or_default();
+        let best_number = blockchain_info.best_number;
```

Comment:
I would split off at least this optimistic block request into a separate function, the logic is completely independent of the one for children and slightly unexpected, so it would be better to parse it in isolation.

Reply:
Oh, and then ofc call that method here, having a `request_targets` method is great.

Reply:
Not sure if I understand correctly: you want sth like

```suggestion
 request_targets() -> Vec<H> {
        let mut targets = children();
        targets.push(top_wanted(number));
        targets
}
```

?

Reply:
I would split this into two functions, one for `best_number` and one for `children` (not sure about the names). And after that we could modify status report to report:
* highest block and hash from `best_number` function result
* number of possible alternative branches from `children` function result if `n > 1`

Reply:
That was what I wanted, yes, but see other comment about possibly dropping `top_wanted` anyway?

If we keep it, then `children` could return a `HashSet`, so we don't need to bother with the `contains` check, and this method in general could return an `IntoIter<Item = H>`.

Reply:
Changed the logic due to request status, not sure how to split it now.

Reply:
Lines 238 to 255 in one method (and really make it return a `HashSet` plx, especially since you `collect` anyway), lines 256 to 275 in another (moreorless, you want to return not just push :P). I wouldn't worry about double-counting requests for a child in the request status, but if you want to that is also pretty simple, since `insert` returns whether the value was already there.

Reply:
Huh, it inspired me to further simplify the current version. Check it out ;)